app/src/main/java/com/sdp13epfl2021/projmag/database/FirebaseFileDatabase.kt

```diff
@@ -1,103 +1,133 @@
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import java.io.File
+import java.io.IOException
+import java.util.*
+
+class FirebaseFileDatabase(
+    private val storage: FirebaseStorage,
+    private val auth: FirebaseAuth
+) : FileDatabase {
+
+    /**
+     * Reference of the root of users in Firebase Storage
+     */
+    private val rootRef = storage.reference.child("users")
+
+
+    override fun getFile(
+        fileUrl: String,
+        destinationFolder: File,
+        onSuccess: (File) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        val fileRef: StorageReference
+        val destinationFile: File
+
+        try {
+            if (!destinationFolder.mkdirs() && !destinationFolder.isDirectory) {
+                throw IOException("The destination folder \"${destinationFolder.path}\" can't be created.")
+            }
+            fileRef = storage.getReferenceFromUrl(fileUrl)
+            destinationFile = File(destinationFolder, fileRef.name)
+        } catch (e: Exception) {
+            GlobalScope.launch { onFailure(e) }
+            return
+        }
+
+        if (destinationFile.exists()) {
+            GlobalScope.launch { onSuccess(destinationFile) }
+            return
+        }
+
+        fileRef
+            .getFile(destinationFile)
+            .addOnSuccessListener { onSuccess(destinationFile) }
+            .addOnFailureListener(onFailure)
+    }
+
+
+    override fun pushFile(
+        file: File,
+        onSuccess: (Uri) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+        if (!file.exists() || !file.isFile) {
+            GlobalScope.launch { onFailure(IOException("The file \"${file.path}\" can't be read.")) }
+            return
+        }
+
+        val user = auth.currentUser
+        if (user == null) {
+            GlobalScope.launch { onFailure(SecurityException("Pushing file can only be done by authenticated user.")) }
+            return
+        }
+
+        val fileRef = rootRef
+            .child(user.uid)
+            .child("${UUID.randomUUID()}_${file.name}")
+
+        fileRef
+            .putFile(file.toUri())
+            .continueWithTask { task ->
+                if (!task.isSuccessful) {
+                    task.exception?.let {
+                        throw it
+                    }
+                }
+                fileRef.downloadUrl
+            }
+            .addOnSuccessListener(onSuccess)
+            .addOnFailureListener(onFailure)
+    }
+
+    override fun pushFileFromUri(
+        uri: Uri,
+        onSuccess: (Uri) -> Unit,
+        onFailure: (Exception) -> Unit
+    ) {
+
+        val user = auth.currentUser
+        if (user == null) {
+            GlobalScope.launch { onFailure(SecurityException("Pushing file can only be done by authenticated user.")) }
+            return
+        }
+
+        val fileRef = rootRef
+            .child(user.uid)
+            .child("${UUID.randomUUID()}_${uri.toString()}")
```

Comment:
Using uri.toString() give a very long string with too many sub directory/etc
For example this video is stored in 7 sub directories below the /users/{userId} directory.
https://firebasestorage.googleapis.com/v0/b/projmag.appspot.com/o/users%2FWkanpcGw5Dhp9J8uvLAIDQ8XhxA3%2F4dec9a82-f45c-42ab-b822-a56ae1b825ad_content%3A%2Fcom.google.android.apps.photos.contentprovider%2F-1%2F2%2Fcontent%253A%252F%252Fmedia%252Fexternal%252Fvideo%252Fmedia%252F28%2FORIGINAL%2FNONE%2F456595140?alt=media&token=92663bb8-b831-40c7-a184-73b40d9e440c

Reply:
What  do you suggest instead, just the UUID

Reply:
Maybe "UUID_originalFilename", or juste the UUID if this is not possible